semgrep-core/Analyzing/Bloom_annotation.ml

```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
  * See also Rules_filter.reserved_id and reserved_str and
  * Rules_filter.extract_specific_strings.
 *)
-let extract_strings _any =
-  []
+
+let rec statement_strings stmt =
+  let res = ref [] in
+  let top_level = ref true in
+  let visitor = V.mk_visitor {
+    V.default_visitor with
+    V.kident = (fun (_k, _) (str, _tok) ->
+      if not (special_ident str) then
+        push str res
+    );
+    V.kexpr = (fun (k, _) x ->
+      (match x with
+       (* less: we could extract strings for the other literals too?
+        * atoms, chars, even int?
+       *)
+       | L (String (str, _tok)) ->
+           if not (special_literal str) then push str res
+       | TypedMetavar _ -> ()
+       | _ -> k x
+      )
+    );
+    V.kconstness = (fun (k, _) x ->
+      (match x with
+       | Lit (String (str, _tok)) ->
+           if not (special_literal str) then push str res
+       | _ -> k x
+      )
+    );
+    V.kid_info = (fun (k, _) x -> k x);
+    V.kstmt = (fun (k, _) x ->
+      (* First statement visited is the current statement *)
+      if !top_level then (top_level := false; k x)
+      else
+        (* For any other statement, recurse to add the filter *)
+        (let strs = statement_strings x in
+         let bf = B.create () in
+         add_all_to_bloom strs bf;
+         push_list strs res;
+         x.s_bf <- Some (bf))
+    );
+  } in
+  visitor (S stmt);
+  !res
+;;
 
 (*****************************************************************************)
 (* Analyze the pattern *)
 (*****************************************************************************)
 
-let bloom_of_expr _e =
-  B.create ()
-
-let bloom_of_stmt _st =
-  B.create ()
+let list_of_pattern_strings any =
+  let res = ref [] in
+  let visitor = V.mk_visitor {
+    V.default_visitor with
+    V.kident = (fun (_k, _) (str, _tok) ->
+      if not (special_ident str) then
+        push str res
+    );
+    V.kexpr = (fun (k, _) x ->
+      (match x with
+       (* less: we could extract strings for the other literals too?
+        * atoms, chars, even int?
+       *)
+       | L (String (str, _tok)) ->
+           if not (special_literal str) then
+             push str res
+       | TypedMetavar _ -> ()
+       | _ -> k x
+      )
+    );
+  } in
+  visitor any;
+  !res
+;;
 
 (*****************************************************************************)
 (* Analyze the code *)
 (*****************************************************************************)
 
 (* TODO: visit AST and set the s_bf field in statements *)
-let annotate_program _ast =
-  ()
+let annotate_program ast =
+  let visitor = V.mk_visitor {
+    V.default_visitor with
+    V.kstmt = (fun (_k, _) x ->
+      let bf = B.create () in
+      let ids = statement_strings x in
```

Comment:
This seems like duplicated with the statement_strings code, why not factorize? (which would also
make useless the is_toplevel variable).

Reply:
See earlier comment (the need to union all the child statements' strings and add it to the statement's bloom filter). There may be a better way of doing this (perhaps passing in a ref list of bloom filters?), but I'm not sure